### PR TITLE
Closes #6157: Improve Pressidium compatibility

### LINKED ISSUE #6157: Improve Pressidium hosting compatibility
**Before submitting an issue please check that you’ve completed the following steps:**
- Made sure you’re on the latest version
- Used the search feature to ensure that the bug hasn’t been reported before

**Describe the bug**
Our current compatibility covers only the following cases:
1. Clearing our whole cache when their cache is purged
2. Clearing their whole cache when our is purged
3. Hide Varnish Config fields

**Expected behavior**
It's possible to clear only the specific URL(s) too, we should handle that.

**Screenshots**
[Documentation](https://github.com/wp-media/wp-rocket/files/12576277/Integrating.with.Pressidium.Manually.invalidating.cache.pdf)

**Acceptance Criteria (for WP Media team use only)**
1. Their cache should be cleared when using `Purge this URL` option from the top menu
2. Their cache should be cleared when updating the post (post itself + related URLs)
3. No regression compared to the current compatibility described in the **Describe the bug** section
4. Specific URLs should be cleared when RUCSS data is injected

### PR DESCRIPTION
**Pull request by Miraeld**
## Description

Restructure of the Pressidium compatibility to be with the new ThirdParty structure.
Improvement of the compatibility with Pressidium. 

Fixes #6157

## Type of change

- Enhancement (non-breaking change which improves an existing functionality).

## Is the solution different from the one proposed during the grooming?

No

# Checklists

## Generic development checklist

- [x] My code follows the style guidelines of this project, with adapted comments and without new warnings.
- [x] I have added unit and integration tests that prove my fix is effective or that my feature works.
- [x] The CI passes locally with my changes (including unit tests, integration tests, linter).
- [ ] Any dependent changes have been merged and published in downstream modules.
- [ ] If applicable, I have made corresponding changes to the documentation. *Provide a link to the documentation.*

## Test summary

- [] I triggered all changed lines of code at least once without new errors/warnings/notices.
- [ ] I validated all Acceptance Criteria of the related issues. (If applicable, provide proof).
- [ ] I validated all test plan the QA Review asked me to.

### CODE REVIEW

File: inc/ThirdParty/Hostings/Pressidium.php [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+	 *
+	 * @return void
+	 */
+	public function clean_post( $post, $purge_urls ) {
+		$cache = NinukisCaching::get_instance();
+		$cache->purge_page_cache( $post->ID );
+
+		// Purge related urls.
+		$paths = $this->get_paths( $purge_urls );
+		$cache->purge_cache( $paths );
```
Comment (remyperona) [inline inc/ThirdParty/Hostings/Pressidium.php]: I don't think the first part is needed, the post URL is already included in the `$purge_urls` array

**Review by remyperona — CHANGES_REQUESTED** [review]

File: inc/ThirdParty/Hostings/Pressidium.php (hunk 1 above)
Comment (Miraeld) [reply inc/ThirdParty/Hostings/Pressidium.php]: It's been taken off.

File: inc/ThirdParty/Hostings/Pressidium.php [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+	/**
+	 * Purge the cache for the given URL.
+	 *
+	 * @param string|array $url URL we want to purge.
+	 *
+	 * @return void
+	 */
+	public function purge_url( $url ) {
+		$paths = $this->get_paths( $url );
+		NinukisCaching::get_instance()->purge_cache( $paths );
```
Comment (remyperona) [inline inc/ThirdParty/Hostings/Pressidium.php]: the code in this method and the next one could be refactored into a shared private method

**Review by remyperona — CHANGES_REQUESTED** [review]

File: inc/ThirdParty/Hostings/Pressidium.php (hunk 2 above)
Comment (Miraeld) [reply inc/ThirdParty/Hostings/Pressidium.php]: It's been changed.

**Review by remyperona — APPROVED** [review]

**Review by vmanthos — APPROVED** [review]
This is working as expected. :+1: 

**TestRail report:** [testrail-report-534.pdf](https://github.com/wp-media/wp-rocket/files/12999528/testrail-report-534.pdf)

**NOTE**

The following error is popping up in the logs but it's not caused by WP Rocket:
```
Oct 18 06:30:37 192.168.128.60 apache2[3360293]: [u4078] [wprocket01.onpressidium.com] [6998449f83a7a137] [104.28.209.174] [error] [client 192.168.134.96:45390] AH01071: Got error 'PHP message: wp-ninukis : requesting page cache purge for hosts [wprocket01.onpressidium.com]'
```
It's there when purging all caches using Pressidium's plugin even when WP Rocket is deactivated.

**Comment by cpaloukas** [comment]
@vmanthos thanks for all your efforts and for bringing the error that appears in the logs to our attention!

**Pull request merged**